### PR TITLE
Add Webhook feature and api to get resources

### PR DESCRIPTION
**Pull request by danmaz74**
* WebhookConfiguration + WebhookCallJob allow calling an external system when something happens
* Api::ResourcesController is authenticated with HMAC and allows to get a list of all resources, or all resources which have a link with a timestamp >= a given value

### CODE REVIEW

File: config/initializers/api.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# frozen_string_literal: true
+
+raise 'API_SECRET_KEY environment variable must be set' unless ENV['API_SECRET_KEY'].present? || Rails.env.test?
```
Comment (shlag3n) [inline config/initializers/api.rb:3]: don't think we need to raise it here, better to check on API call level, not every lcms needs API.

File: app/controllers/lcms/engine/api/resources_controller.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+module Lcms
+  module Engine
+    module Api
+      class ResourcesController < BaseController
+        def index
+          resources = Resource
```
Comment (shlag3n) [inline app/controllers/lcms/engine/api/resources_controller.rb:8]: we can move `.all` here

File: app/controllers/lcms/engine/api/resources_controller.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+          # format for the query string is: ?link_updated_after={link_path}:{timestamp}
+          if params[:link_updated_after]
+            link_path, timestamp = params[:link_updated_after].split(':')
+
+            resources = resources.where_link_updated_after(link_path, Time.at(timestamp.to_i))
+          end
+
+          render json: resources.all.as_json
```
Comment (shlag3n) [inline app/controllers/lcms/engine/api/resources_controller.rb:17]: Do we need special serializer for that?

File: app/models/lcms/engine/integrations/webhook_configuration.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        def execute_call(payload)
+          response = HTTParty.send(
+            action.downcase.to_sym,
+            endpoint_url,
+            body: payload.to_json,
+            headers: {
+              'Content-Type' => 'application/json'
+            }.merge(auth_headers(payload)),
+            timeout: 30
```
Comment (shlag3n) [inline app/models/lcms/engine/integrations/webhook_configuration.rb:30]: better to move to constant

File: app/jobs/lcms/engine/integrations/webhook_call_job.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+module Lcms
+  module Engine
+    module Integrations
+      class WebhookCallJob < Lcms::Engine::ApplicationJob
+        extend ResqueJob
+        include Lcms::Engine::RetryDelayed
+
+        queue_as :low
```
Comment (shlag3n) [inline app/jobs/lcms/engine/integrations/webhook_call_job.rb:10]: need to mention at docs that we're using `low` queue, we may not have it.

File: config/initializers/api.rb (hunk 1 above)
Comment (shlag3n) [reply config/initializers/api.rb:3]: Also we need to add it to `.env.template` + to https://github.com/learningtapestry/lcms-engine/blob/master/docs/env-variables.md

File: app/controllers/lcms/engine/api/base_controller.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+          timestamp = request.headers['X-Api-Timestamp']
+          signature = request.headers['X-Api-Signature']
+
+          unless timestamp && signature
+            render json: { error: 'Missing authentication headers' }, status: :unauthorized
+            return
+          end
+
+          # Verify timestamp is within 5 minutes
+          if (Time.now.to_i - timestamp.to_i).abs > 300
```
Comment (shlag3n) [inline app/controllers/lcms/engine/api/base_controller.rb:23]: better to move `300` to constant

Comment (shlag3n) [reply app/controllers/lcms/engine/api/base_controller.rb:23]: and we may need to add `skip_before_action :verify_authenticity_token`

File: config/initializers/api.rb (hunk 1 above)
Comment (danmaz74) [reply config/initializers/api.rb:3]: Done

File: app/controllers/lcms/engine/api/resources_controller.rb (hunk 3 above)
Comment (danmaz74) [reply app/controllers/lcms/engine/api/resources_controller.rb:17]: I don't think so

File: app/jobs/lcms/engine/integrations/webhook_call_job.rb (hunk 5 above)
Comment (danmaz74) [reply app/jobs/lcms/engine/integrations/webhook_call_job.rb:10]: I had seen it here:

```
    class DocumentBundleGenerateJob < Lcms::Engine::ApplicationJob
      extend ResqueJob

      queue_as :low
```

So I suppose it's already set up?

File: app/controllers/lcms/engine/api/base_controller.rb (hunk 6 above)
Comment (danmaz74) [reply app/controllers/lcms/engine/api/base_controller.rb:23]: Added the constant. skip_before_action :verify_authenticity_token shouldn't be necessary when using protect_from_forgery with: :null_session

File: app/models/lcms/engine/integrations/webhook_configuration.rb (hunk 4 above)
Comment (danmaz74) [reply app/models/lcms/engine/integrations/webhook_configuration.rb:30]: Done

File: app/controllers/lcms/engine/api/resources_controller.rb (hunk 2 above)
Comment (danmaz74) [reply app/controllers/lcms/engine/api/resources_controller.rb:8]: Why?

Comment (shlag3n) [reply app/controllers/lcms/engine/api/resources_controller.rb:8]: There will be `.all` for `.where` too. Anyway up to you, it's mostly just preference just pass ready result to json serializer.

File: app/jobs/lcms/engine/integrations/webhook_call_job.rb (hunk 5 above)
Comment (shlag3n) [reply app/jobs/lcms/engine/integrations/webhook_call_job.rb:10]: If we use that job yes, but if not - no, imo it's just good to know that we're using `low` queue without checking the code.

File: app/controllers/lcms/engine/api/resources_controller.rb (hunk 3 above)
Comment (shlag3n) [reply app/controllers/lcms/engine/api/resources_controller.rb:17]: Ok, just thought maybe we don't need to fetch all fields we have now at `ResourceSerializer` for this request but have `links` field.

**Review by shlag3n — APPROVED** [review]

File: db/migrate/20241024102540_create_lcms_engine_integrations_webhook_configurations.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+class CreateLcmsEngineIntegrationsWebhookConfigurations < ActiveRecord::Migration[6.1]
+  def change
+    create_table :lcms_engine_integrations_webhook_configurations do |t|
+      t.string :event_name, null: false, index: { name: 'index_webhook_configurations_on_event_name' }
+      t.boolean :active, default: true
```
Comment (paranoicsan) [inline db/migrate/20241024102540_create_lcms_engine_integrations_webhook_configurations.rb:7]: I would add `null: false` as PostgreSQL can have `true|false|null` values in boolean fields. But that's not a critical for now I suppose.

**Review by paranoicsan — APPROVED** [review]
Looks awesome! Thanks @danmaz74!

**Comment by paranoicsan** [comment]
@danmaz74 I think can be logically squashed. And I will merge.

File: app/jobs/lcms/engine/integrations/webhook_call_job.rb (hunk 5 above)
Comment (mbruszewski) [reply app/jobs/lcms/engine/integrations/webhook_call_job.rb:10]: We tested that and it was working with low. Resolving.

File: db/migrate/20241024102540_create_lcms_engine_integrations_webhook_configurations.rb (hunk 7 above)
Comment (mbruszewski) [reply db/migrate/20241024102540_create_lcms_engine_integrations_webhook_configurations.rb]: thanks, added

**Pull request merged**